syntax/grammar.mjs

```diff
@@ -0,0 +1,502 @@
+/* Tokens */
+let identifier =
+    sequence(
+        charset("a-zA-Z"),
+        repeat(
+            charset("a-zA-Z0-9_-")))
+    .map(flatten(1))
+    .map(join);
+
+let comment_line = defer(() =>
+    either(
+        sequence(
+            line_end.abstract),
+        sequence(
+            char(" "),
+            regex(/.*/).abstract,
+            line_end.abstract))
+    .map(keep_abstract)
+    .map(join));
+
+let word = defer(() =>
+    repeat1(
+        and(
+            not(char("[")),
+            not(char("]")),
+            not(char("{")),
+            not(char("}")),
+            not(backslash),
+            regular_char))
+    .map(join));
+
+/* ---------- */
+/* Characters */
+
+let backslash = char("\\");
+let quote = char("\"");
+
+/* Any Unicode character from BMP excluding C0 control characters, space,
+ * surrogate blocks and non-characters (U+FFFE, U+FFFF).
+ * Cf. https://www.w3.org/TR/REC-xml/#NT-Char
+ * TODO Add characters from other planes: U+10000 to U+10FFFF.
+ */
+let regular_char =
+    charset("\u0021-\uD7FF\uE000-\uFFFD");
+
+let text_char = defer(() =>
+    either(
+        inline_space,
+        regex(/\\u[0-9a-fA-F]{4}/),
+        sequence(
+            backslash,
+            backslash).map(join),
+        sequence(
+            backslash,
+            char("{")).map(join),
+        and(
+            not(backslash),
+            not(char("{")),
+            regular_char)));
+
+let text_cont = defer(() =>
+    sequence(
+        break_indent,
+        and(
+            not(char(".")),
+            not(char("*")),
+            not(char("[")),
+            not(char("}")),
+            text_char))
+    .map(join));
+
+let quoted_text_char =
+    either(
+        and(
+            not(quote),
+            text_char),
+        sequence(
+            backslash,
+            quote).map(join));
+
+let quoted_text =
+    sequence(
+        quote,
+        repeat(quoted_text_char),
+        quote)
+    .map(element_at(1))
+    .map(join);
+
+let digit = charset("0-9");
+
+let number =
+    sequence(
+        maybe(char("-")),
+        repeat1(digit),
+        maybe(
+            sequence(
+                char("."),
+                repeat1(digit))))
+    .map(flatten(2))
+    .map(join);
+
+/* ---------- */
+/* Whitespace */
+let inline_space =
+    repeat1(
+        either(
+            char("\u0020"),
+            char("\u0009")))
+    .map(join);
+
+let line_end =
+    either(
+        string("\u000D\u000A"),
+        char("\u000A"),
+        char("\u000D"),
+        eof());
+
+let blank_line =
+    sequence(
+        maybe(inline_space),
+        line_end)
+    .map(join);
+
+let break_indent =
+    sequence(
+        line_end,
+        repeat(blank_line),
+        inline_space)
+    .map(flatten(1))
+    .map(join);
+
+let space_indent =
+    either(
+        sequence(
+            inline_space,
+            maybe(break_indent)),
+        sequence(
+            maybe(inline_space),
+            break_indent));
```

Comment:
I'd like to keep using the `maybe()` combinator because it serializes as `?` in the EBNF. Your version of `space_indent` is how I originally implemented it :) I changed it to remove the redyndancy of nested `maybes`. If you're OK with it, I'll go with `maybe(space_indent)` and your implementation of `space_indent`.